### ISSUE TITLE
Split a layer into multiple layers by drawing distance

### ISSUE THREAD

**Issue by cbmoore** [abey79/vpype#337]
It would be helpful to be able to split a layer into multiple layers by distance drawn.  The use case is so that if i know my pen needs a refill (or pencil needs sharpening) after 50 meters drawn, I can have the plotter plot one layer and then wait for me to refill/sharpen before starting the next layer.

**Comment by abey79** [comment]
Sounds like a good idea, and rather easy to implement, thanks for suggesting.

**Comment by andeecollard** [comment]
Can I second this request! I mentioned this several months ago on the Drawbots Discord and haven't really got much further. I would love something that replicates the functionally of the Watercolorbot Inkscape extension. Having fixed paint palette locations that correlate to layer colours and for the brush to automatically replenish based on distance would be amazing!

I think it would work particularly well with grbl machines that can't use the watercolorbot extension.

Would vsketch be the best place for me to try and implement this?

**Comment by abey79** [comment]
vpype is the right place for this.

Spec (TBD):
- command name `splitdist` with a `dist` argument
- it would be a global_processor, albeit with a `--layer` option
- the original layers would keep the n first lines where their cumulated length is `<= dist`
- line length: `np.cumsum(np.abs(np.diff(line)))`
- surplus lines would be spread in new layers, taking the lowest available IDs (`vp.Document.free_id()`)
- newly created layers would have their metadata copied from the original layer (`vp.LineCollection.clone()`)

@LoicGoulefert any motivation for a PR? ;)

Though it would be possible, it'd be tricky to implement this in vsketch since you don't now until after the fact the length of a given geometry (due to transforms). You *could* check the total length of the lines in `vsk._document.layers[cur_layer]` after each draw call and then call `vsk.stroke(cur_layer := cur_layer + 1)` when it goes over a threshold, but that would be inconvenient and ugly.

**Comment by andeecollard** [comment]
@abey79 thanks— I’ll do some research

**Comment by abey79** [comment]
Fixed in #487

@andeecollard Could you maybe give it a shot before I release 3.11?

**Comment by cbmoore** [comment]
@abey79 I could give it a try.  I'm running on an M1 Mac with vpype installed by the recommended method in MacPorts.   Is there a reasonably easy way to install the head of the main branch as the current version?

**Comment by cbmoore** [comment]
Thanks for the instructions.  I have installed it and splitdist appears
to be working correctly based on the appearance of the output in Inkscape.
The only hitch I ran into is in the specification of the distance.  I'm
using a Pentel Orenz Nero pencil that I need to refill after 130m of line
drawn but "m" is not a valid unit in vpype.  "13000cm" works fine but is
prone to error so I'd suggest adding meters to the allowed units.

Thank you for this, it's going to make pencil drawings much easier!

**Comment by jimmykl** [comment]
> 



> Pentel Orenz Nero pencil

Excuse the off topic question but does the auto-advance on these pencils work well with a plotter?

**Comment by cbmoore** [comment]
@jimmykl Yes.  I'm using an AxiDraw with 30g of added weight.  It works perfectly except for the last ~1cm of lead which cannot be automatically advanced (not a plotter problem, just how the pencil works).  You then have to manually advance the next lead into position.

**Comment by abey79** [comment]
@cbmoore Excellent point about adding meters. I'll add them as well as feet.

**Comment by abey79** [comment]
Added `m` and `ft` in #498

**Comment by andeecollard** [comment]
> Fixed in #487
> 
> @andeecollard Could you maybe give it a shot before I release 3.11?

This is brilliant! The only thing I'm missing is the reload routine. I could probably bodge together something manually with the svg but this obviously isn't ideal

**Comment by cbmoore** [comment]
@andeecollard I'm reloading the pencil by hand so my use case is simple.  I run something like vpype read <infile> splitdist 130m write <outfile>.  This separates the lines into separate layers of <= 130m line length.  I plot one, reload the pencil, and then plot the next one.

**Comment by cbmoore** [comment]
@abey79 
> Edit: any example of such a pencil plot out there?

I haven't taken and pictures yet--coming soon.

**Comment by andeecollard** [comment]
I’ll try!
